gix-dir/src/walk/readdir.rs

```diff
@@ -59,12 +59,24 @@ pub(super) fn recursive(
         );
         current.push(file_name);
 
+        let disk_kind = {
+            let m = current.metadata().map_err(|err| Error::DirEntryFileType {
+                path: std::mem::take(current),
+                source: err,
+            })?;
+            match Kind::try_from(m.file_type()) {
+                Ok(kind) => kind,
+                Err(_err) => {
+                    continue;
```

Comment:
> I am surprised these seemingly sweeping changes are needed when trying to ignored 'special' files

Yeah they're not, the only change that's needed is this "continue", which should probably be its own commit.

But I believe other places can be simplified and made more consistent. I may be wrong, I just need to find time to finish it

Reply:
One will definitely need tests for that as well, and in the crates you are working in, they are basically done in a test-driven fashion exclusively as the code is way to complex to understand exactly what's going to happen.
But with the re-organization of commits I think all this will come quite naturally.